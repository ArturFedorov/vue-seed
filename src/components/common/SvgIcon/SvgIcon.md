Adds an SVG-icon from sprite created by 'vue-cli-plugin-svg-sprite' plugin.

___


#### Examples of the component
```jsx
<SvgIcon
  name="tick"
  width="30"
  height="30"
  title="tick" />

<SvgIcon
  name="tick"
  width="30"
  height="30"
  title="tick"
  class="dark-icon" />

<SvgIcon
  name="tick"
  width="30"
  height="30"
  title="tick"
  class="brand-icon" />

<SvgIcon
  name="tick"
  width="30"
  height="30"
  title="tick"
  class="gray-icon" />

<span style="background:gray;display:inline-block;padding:4px;">

<SvgIcon
  name="tick"
  width="30"
  height="30"
  title="tick"
  class="white-icon" />

</span>
```
