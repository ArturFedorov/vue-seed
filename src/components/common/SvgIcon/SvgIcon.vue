<template>
  <svg :class="className" xmlns="http://www.w3.org/2000/svg" :width="width" :height="height">
    <title v-if="title">{{ title }}</title>
    <use :href="iconPath"/>
  </svg>
</template>

<script lang="ts">
import Vue from 'vue';

export default Vue.extend({
  name: 'SvgIcon',

  props: {
    name: {
      type: String,
      required: true
    },

    title: {
      type: String,
      default: ''
    },

    width: {
      type: [String, Number],
      required: false
    },

    height: {
      type: [String, Number],
      required: false
    }
  },

  computed: {
    iconPath(): string {
      try {
        return require(`@/assets/icons/${this.name}.svg`).default.url;
      } catch (ex) {
        return '';
      }
    },

    className(): string {
      return 'svg-icon svg-icon-' + this.name;
    }
  },

  watch: {
    name() {
      /*
      * We need this because otherwise shadowDOM isn't updated
      */
      this.$el.innerHTML = `<use xlink:href="${this.iconPath}"/>`;
    }
  }
});
</script>

<style lang="scss" scoped>
  .svg-icon {
    transition: fill .3s ease-in-out;
  }
</style>
