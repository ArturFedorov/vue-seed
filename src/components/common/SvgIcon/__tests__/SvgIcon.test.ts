import { shallowMount, mount } from '@vue/test-utils';
import SvgIcon from '../SvgIcon.vue';

describe('SvgIcon.vue', () => {
  const selectors = {
    title: 'title'
  };

  const defaultPropsData = {
    name: 'tick',
    width: '10',
    height: '10',
    title: 'tick'
  };

  function mountComponent(propsData = defaultPropsData as any) {
    return shallowMount(SvgIcon, {
      propsData
    });
  }

  describe('Adds attributes and values according to props', () => {
    test('sets string props correctly', () => {
      const wrapper = mountComponent();
      expect(wrapper.find(selectors.title).exists()).toBe(true);
      expect(wrapper.find(selectors.title).text()).toBe('tick');
      expect(wrapper.attributes().width).toBe('10');
      expect(wrapper.attributes().height).toBe('10');
    });

    test('sets number props correctly', () => {
      const wrapper = mountComponent({
        name: 'tick',
        width: 10,
        height: 10
      });
      expect(wrapper.attributes().width).toBe('10');
      expect(wrapper.attributes().height).toBe('10');
    });
  });

  test('Renders correctly', () => {
    const wrapper = mountComponent();
    expect(wrapper.element).toMatchSnapshot();
  });
});
