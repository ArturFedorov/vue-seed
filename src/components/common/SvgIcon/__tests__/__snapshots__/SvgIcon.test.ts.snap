// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SvgIcon.vue Renders correctly 1`] = `
<svg
  class="svg-icon svg-icon-tick"
  height="10"
  width="10"
  xmlns="http://www.w3.org/2000/svg"
>
  <title>
    tick
  </title>
   
  <use
    href=""
  />
</svg>
`;
